presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java

```diff
@@ -236,25 +238,48 @@ final boolean internalExecute(String sql)
 
         StatementClient client = null;
         PrestoResultSet resultSet = null;
+        boolean intercepted = false;
+
         try {
-            client = connection().startQuery(sql, getStatementSessionProperties());
-            if (client.isFinished()) {
-                QueryStatusInfo finalStatusInfo = client.finalStatusInfo();
-                if (finalStatusInfo.getError() != null) {
-                    throw resultsException(finalStatusInfo);
-                }
-            }
-            executingClient.set(client);
             WarningsManager warningsManager = new WarningsManager();
             currentWarningsManager.set(Optional.of(warningsManager));
-            resultSet = new PrestoResultSet(this, client, maxRows.get(), progressConsumer, warningsManager);
 
-            for (Map.Entry<String, SelectedRole> entry : client.getSetRoles().entrySet()) {
-                connection.get().setRole(entry.getKey(), entry.getValue());
+            int statementDepth = this.statementDepth.incrementAndGet();
+            boolean shouldIntercept = !connection().getQueryInterceptorInstances().isEmpty() && statementDepth == 1;
+
+            if (shouldIntercept) {
+                resultSet = connection().invokeQueryInterceptorsPre(sql, this);
+            }
+
+            // Check if no resultSet is returned from an interceptor
+            if (resultSet != null) {
+                currentResult.set(resultSet);
+                intercepted = true;
+            }
+            else {
+                client = connection().startQuery(sql, getStatementSessionProperties());
+                if (client.isFinished()) {
+                    QueryStatusInfo finalStatusInfo = client.finalStatusInfo();
+                    if (finalStatusInfo.getError() != null) {
+                        throw resultsException(finalStatusInfo);
+                    }
+                }
+                executingClient.set(client);
+
+                resultSet = new PrestoResultSet(this, client, maxRows.get(), progressConsumer, warningsManager);
+
+                for (Map.Entry<String, SelectedRole> entry : client.getSetRoles().entrySet()) {
+                    connection.get().setRole(entry.getKey(), entry.getValue());
+                }
             }
 
             // check if this is a query
-            if (client.currentStatusInfo().getUpdateType() == null) {
+            if (intercepted || client.currentStatusInfo().getUpdateType() == null) {
+                currentResult.set(resultSet);
```

Comment:
Previously (without innerResultSet logic), if an interceptor's postProcess() executes a query using the intercepted Statement and returns a ResultSet then we'll lose the reference to the top level query's ResultSet and it will stay opened past the lifetime of the Statement. Adding it here means it can be cleared in clearCurrentResults() when the interceptor's postProcess() executes another query.

Reply:
Makes sense ! Should we move the second `set()` on line 283 to be inside the `if` condition then ?

Reply:
Ah of course